buildSrc/src/main/groovy/io.deephaven.java-conventions.gradle

```diff
@@ -74,18 +76,85 @@ tasks.withType(JavaCompile).configureEach {
   }
 }
 
+def tier4CompilationExcludes = [
+  'io.deephaven.engine.table.impl.by.ssmpercentile.DoublePercentileTypeMedianHelper':['setResult'],
+  'io.deephaven.engine.table.impl.by.ssmpercentile.FloatPercentileTypeMedianHelper':['setResult'],
+  'io.deephaven.engine.table.impl.by.ssmpercentile.SsmChunkedPercentileOperator':['addChunk'],
+  'io.deephaven.engine.table.impl.by.IncrementalChunkedOperatorAggregationStateManager':['buildTable'],
+  'io.deephaven.engine.table.impl.RightIncrementalChunkedCrossJoinStateManager':['addToIndex', 'decorationProbe'],
+  'io.deephaven.engine.table.impl.IncrementalChunkedNaturalJoinStateManager':['buildTable', 'decorationProbe'],
+]
+def createCompilerDirectives = tasks.register('createCompilerDirectives') {
+  def c2ExcludesFile = project.layout.buildDirectory.file('c2-excludes.txt')
+  it.outputs.file(c2ExcludesFile)
+
+  doFirst {
+    def excludesAsList = []
+    tier4CompilationExcludes.forEach({type, methods ->
+      methods.forEach {method ->
+        excludesAsList += "${type}::${method}"
+      }
+    })
+    def builder = new JsonBuilder()
+    builder([{
+              match excludesAsList
+              c2 {
+                Exclude true
+              }
+            }])
+    c2ExcludesFile.get().asFile.text = builder.toPrettyString()
+  }
+}
+
+def c2ExcludesJvmArgs = { String c2ExcludesFile ->
+  return [
+          '-XX:+UnlockDiagnosticVMOptions',
+          "-XX:CompilerDirectivesFile=${c2ExcludesFile}",
+  ]
+}
+
+// Utility to add jvm args to all executions, whether intellij or from a application script or gradle javaexec
+def devJvmArgs = [
+//  '-XX:Tier4CompileThreshold=1000', // this optional line makes it easier to trigger the c2 error on the above methods
+//  '-XX:+PrintFlagsFinal',           // this optional line makes it easier to explore the final values for compiler args
+//  '-XX:+PrintCompilation',          // this optional line shows jit operations as they happen
+]
+
 tasks.withType(JavaExec).configureEach {
+  def c2ExcludesFile = createCompilerDirectives.get().outputs.files
+  inputs.files c2ExcludesFile
   javaLauncher.set runtimeLauncher
+  jvmArgs += c2ExcludesJvmArgs(c2ExcludesFile.singleFile.path) + devJvmArgs
 }
 
 tasks.withType(Test).configureEach {
+  def c2ExcludesFile = createCompilerDirectives.get().outputs.files
+  inputs.files c2ExcludesFile
+
   javaLauncher.set testRuntimeLauncher
+  jvmArgs += c2ExcludesJvmArgs(c2ExcludesFile.singleFile.path) + devJvmArgs
 }
 
 tasks.withType(GroovyCompile).configureEach {
   javaLauncher.set groovyCompilerLauncher
 }
 
+plugins.withType(ApplicationPlugin) {
+  applicationDistribution.into('lib') {
```

Comment:
Do we have strong opinions about where the file should live?

I could argue for the directory `lib`, `files` (or some other generic name), or even into the root directory.

Reply:
lib or root dir was my guess, rather than make a new dir. I don't have a strong opinion here, will take any suggestion/advice you offer. At the end of the day, anything in the archive will work, and will be "owned" by the dh-core install.

If I had to make a new dir, it was going to be `etc/`, as a "these are default configs, which you might change, but the defaults are necessary".

Reply:
Do you have a preference, or are you okay with lib/?

Reply:
I'm fine with `lib/` ATM. If we find ourselves needing more special files in the future, I think something like `etc/` may be called for.